native/dispatch.c

```diff
@@ -3346,39 +3364,46 @@ Java_com_sun_jna_Native_getAPIChecksum(JNIEnv *env, jclass UNUSED(classp)) {
   return newJavaString(env, CHECKSUM, CHARSET_UTF8);
 }
 
-JNIEXPORT jint JNICALL
-JNI_OnLoad(JavaVM *jvm, void *UNUSED(reserved)) {
+JNIEXPORT jboolean JNICALL Java_com_sun_jna_Native_isStaticEnabled(JNIEnv * UNUSED(env), jclass UNUSED(classp)) {
+    return JNI_TRUE;
+}
+
+JNIEXPORT jint JNICALL Java_com_sun_jna_Native_initializeStatic(JNIEnv *env, jclass UNUSED(classp)) {
+    return initializeJnaStatics(env);
+}
+
+jint setupJna(JavaVM *jvm) {
   JNIEnv* env;
-  int result = JNI_VERSION_1_4;
   int attached = (*jvm)->GetEnv(jvm, (void *)&env, JNI_VERSION_1_4) == JNI_OK;
-  const char* err;
-
   if (!attached) {
     if ((*jvm)->AttachCurrentThread(jvm, (void *)&env, NULL) != JNI_OK) {
       fprintf(stderr, "JNA: Can't attach native thread to VM on load\n");
       return 0;
     }
   }
-
-  if ((err = JNA_init(env)) != NULL) {
-    fprintf(stderr, "JNA: Problems loading core IDs: %s\n", err);
-    result = 0;
-  }
-  else if ((err = JNA_callback_init(env)) != NULL) {
-    fprintf(stderr, "JNA: Problems loading callback IDs: %s\n", err);
-    result = 0;
-  }
+  int result = initializeJnaStatics(env);
   if (!attached) {
     if ((*jvm)->DetachCurrentThread(jvm) != 0) {
       fprintf(stderr, "JNA: could not detach thread on initial load\n");
     }
   }
-
   return result;
 }
 
-JNIEXPORT void JNICALL
-JNI_OnUnload(JavaVM *vm, void *UNUSED(reserved)) {
+#ifdef JNA_STATIC
+JNIEXPORT jint JNICALL
+JNI_OnLoad_jnidispatch(JavaVM *jvm, void *UNUSED(reserved)) {
+  setupJna(jvm);
+  return JNI_VERSION_1_8;  // upgrade to JNI_VERSION_1_8; required for static JNI
```

Comment:
This is [required by spec](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/invocation.html#JNJI_OnLoad) for static JNI support. `JNI_VERSION_1_8` was introduced in JDK8, so this is safe to apply now that JNA has a minimum that matches.